bug(rapid): unstable MTR cases.

### Search before asking

- [x] I had searched in the [issues](https://github.com/Shannon-Data/ShannonBase/issues) and found no similar issues.


### Version

the laster version.

### What's Wrong?

```
2025-08-12T13:40:28Z UTC - mysqld got signal 6 ;
Most likely, you have hit a bug, but this error can also be caused by malfunctioning hardware.
BuildID[sha1]=346297f9fcbecd6ca0c8ce417322539c580833a0
Thread pointer: 0x7fcba4000fd0
Attempting backtrace. You can use the following information to find out
where mysqld died. If you see no messages after this, something went
terribly wrong...
stack_bottom = 7fcc443a6fe0 thread_stack 0x100000
 #0 0x55e8aa11f250 _ZL18print_fatal_signaliP9siginfo_t
 #1 0x55e8aa11f74c _Z19handle_fatal_signaliP9siginfo_tPv
 #2 0x7fcc4e24532f <unknown>
 #3 0x7fcc4e29eb2c pthread_kill
 #4 0x7fcc4e24527d gsignal
 #5 0x7fcc4e2288fe abort
 #6 0x7fcc4e2297b5 <unknown>
 #7 0x7fcc4e2a8ff4 <unknown>
 #8 0x7fcc4e2ab329 <unknown>
 #9 0x7fcc4e2addad cfree
 #10 0x55e8abb8937c _ZN11ShannonBase4Imcs5Index3ART12Destroy_nodeEPNS2_8Art_nodeE
 #11 0x55e8abbc5b38 _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St10unique_ptrIN11ShannonBase4Imcs5Index5IndexIhmEESt14default_deleteISD_EEESaISH_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSJ_18_Mod_range_hashingENSJ_20_Default_ranged_hashENSJ_20_Prime_rehash_policyENSJ_17_Hashtable_traitsILb1ELb0ELb1EEEE5clearEv
 #12 0x55e8abbc6227 _ZN11ShannonBase4Imcs5TableD0Ev
 #13 0x55e8abb7d893 _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St10unique_ptrIN11ShannonBase4Imcs10RapidTableESt14default_deleteISB_EEESaISF_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSH_18_Mod_range_hashingENSH_20_Default_ranged_hashENSH_20_Prime_rehash_policyENSH_17_Hashtable_traitsILb1ELb0ELb1EEEE8_M_eraseESt17integral_constantIbLb1EERS7_.isra.0
 #14 0x55e8abb823c5 _ZN11ShannonBase4Imcs4Imcs13unload_innodbEPKNS_18Rapid_load_contextEPKcS6_b
 #15 0x55e8abb6c8cd _ZN11ShannonBase8ha_rapid12unload_tableEPKcS2_b
 #16 0x55e8aa02099c _ZL29secondary_engine_unload_tableP3THDPKcS2_RKN2dd5TableEb
 #17 0x55e8aa02ec34 _ZL15drop_base_tableP3THDRK15Drop_tables_ctxP9Table_refbPSt3setIP10handlertonSt4lessIS8_ESaIS8_EEP31Foreign_key_parents_invalidatorPSt6vectorIP10MDL_ticketSaISI_EEP8MEM_ROOT
 #18 0x55e8aa0347a4 _Z23mysql_rm_table_no_locksP3THDP9Table_refbbbPbPSt3setIP10handlertonSt4lessIS6_ESaIS6_EEP31Foreign_key_parents_invalidatorPSt6vectorIP10MDL_ticketSaISG_EE
 #19 0x55e8a9f1f209 _Z11mysql_rm_dbP3THDRK17MYSQL_LEX_CSTRINGb
 #20 0x55e8a9f77806 _Z21mysql_execute_commandP3THDb
 #21 0x55e8a9f7a522 _Z20dispatch_sql_commandP3THDP12Parser_state
 #22 0x55e8a9f7cdbd _Z16dispatch_commandP3THDPK8COM_DATA19enum_server_command
 #23 0x55e8a9f7da1d _Z10do_commandP3THD
 #24 0x55e8aa10e1cf handle_connection
 #25 0x55e8abf3c273 pfs_spawn_thread
 #26 0x7fcc4e29caa3 <unknown>
 #27 0x7fcc4e329c3b <unknown>
 #28 0xffffffffffffffff <unknown>

On X86 Git Action Runner.
```

### How to Reproduce?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

```
[ 71%] binlog.binlog_reset_master_delete_file_failure 'row' w3  [ fail ]
        Test ended at 2025-08-14 02:38:37

CURRENT_TEST: binlog.binlog_reset_master_delete_file_failure
safe_process[244509]: Child process: 244510, killed by signal: 15


Server [mysqld.1 - pid: 244614, winpid: 244614, exit: 256] failed during test run
Server log from this test:
----------SERVER LOG START-----------
Error: -14T02:38:33.009180Z 11 [ERROR] [MY-010831] [Server] Failed to delete file './binlog.000001'
2025-08-14T02:38:33.015493Z 11 [System] [MY-013172] [Server] Received SHUTDOWN from user root. Shutting down mysqld (Version: 8.4.3).
2025-08-14T02:38:33.016108Z 0 [Note] [MY-015023] [Server] MySQL Server: Closing Connections - start.
2025-08-14T02:38:33.016218Z 0 [Note] [MY-010067] [Server] Giving 1 client threads a chance to die gracefully
2025-08-14T02:38:33.016236Z 0 [Note] [MY-010117] [Server] Shutting down replica threads
2025-08-14T02:38:33.016246Z 0 [Note] [MY-010054] [Server] Event Scheduler: Killing the scheduler thread, thread id 5
2025-08-14T02:38:33.016255Z 0 [Note] [MY-010050] [Server] Event Scheduler: Waiting for the scheduler thread to reply
2025-08-14T02:38:33.016376Z 0 [Note] [MY-010048] [Server] Event Scheduler: Stopped
2025-08-14T02:38:33.016389Z 0 [Note] [MY-010118] [Server] Forcefully disconnecting 0 remaining clients
2025-08-14T02:38:33.016398Z 0 [Note] [MY-010043] [Server] Event Scheduler: Purging the queue. 0 events
2025-08-14T02:38:33.016428Z 0 [Note] [MY-015024] [Server] MySQL Server: Closing Connections - end.
2025-08-14T02:38:33.017615Z 0 [Note] [MY-012330] [InnoDB] FTS optimize thread exiting.
2025-08-14T02:38:33.174276Z 0 [Note] [MY-010120] [Server] Binlog end
2025-08-14T02:38:33.174511Z 0 [Note] [MY-015019] [Server] MySQL Server: Plugins Shutdown - start.
2025-08-14T02:38:33.174530Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'mysql_native_password'
2025-08-14T02:38:33.174541Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'mysqlx'
2025-08-14T02:38:33.174972Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'mysqlx_cache_cleaner'
2025-08-14T02:38:33.174990Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'ngram'
2025-08-14T02:38:33.174998Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'BLACKHOLE'
2025-08-14T02:38:33.175006Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'ARCHIVE'
2025-08-14T02:38:33.175012Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'TempTable'
2025-08-14T02:38:33.175018Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'Rapid'
2025-08-14T02:38:33.175241Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'PERFORMANCE_SCHEMA'
2025-08-14T02:38:33.175276Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'MRG_MYISAM'
2025-08-14T02:38:33.175287Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'MyISAM'
2025-08-14T02:38:33.175304Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_SESSION_TEMP_TABLESPACES'
2025-08-14T02:38:33.175312Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_CACHED_INDEXES'
2025-08-14T02:38:33.175318Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_VIRTUAL'
2025-08-14T02:38:33.175324Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_COLUMNS'
2025-08-14T02:38:33.175330Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_TABLESPACES'
2025-08-14T02:38:33.175337Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_INDEXES'
2025-08-14T02:38:33.175343Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_TABLESTATS'
2025-08-14T02:38:33.175350Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_TABLES'
2025-08-14T02:38:33.175356Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_FT_INDEX_TABLE'
2025-08-14T02:38:33.175362Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_FT_INDEX_CACHE'
2025-08-14T02:38:33.175369Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_FT_CONFIG'
2025-08-14T02:38:33.175376Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_FT_BEING_DELETED'
2025-08-14T02:38:33.175383Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_FT_DELETED'
2025-08-14T02:38:33.175389Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_FT_DEFAULT_STOPWORD'
2025-08-14T02:38:33.175411Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_METRICS'
2025-08-14T02:38:33.175420Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_TEMP_TABLE_INFO'
2025-08-14T02:38:33.175426Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_BUFFER_POOL_STATS'
2025-08-14T02:38:33.175432Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_BUFFER_PAGE_LRU'
2025-08-14T02:38:33.175439Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_BUFFER_PAGE'
2025-08-14T02:38:33.175453Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_CMP_PER_INDEX_RESET'
2025-08-14T02:38:33.175459Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_CMP_PER_INDEX'
2025-08-14T02:38:33.175483Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_CMPMEM_RESET'
2025-08-14T02:38:33.175494Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_CMPMEM'
2025-08-14T02:38:33.175520Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_CMP_RESET'
2025-08-14T02:38:33.175528Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_CMP'
2025-08-14T02:38:33.175535Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'INNODB_TRX'
2025-08-14T02:38:33.175542Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'InnoDB'
2025-08-14T02:38:33.175761Z 0 [Note] [MY-013072] [InnoDB] Starting shutdown...
2025-08-14T02:38:33.176352Z 0 [Note] [MY-011944] [InnoDB] Dumping buffer pool(s) to /home/shannon-bin/mysql-test/var/3/mysqld.1/data/ib_buffer_pool
2025-08-14T02:38:33.176582Z 0 [Note] [MY-011944] [InnoDB] Buffer pool(s) dump completed at 250814  5:38:33
2025-08-14T02:38:33.196828Z 0 [Note] [MY-013084] [InnoDB] Log background threads are being closed...
2025-08-14T02:38:33.210461Z 0 [Note] [MY-013854] [InnoDB] Bytes written to disk by DBLWR (ON): 999424
2025-08-14T02:38:33.210669Z 0 [Note] [MY-012980] [InnoDB] Shutdown completed; log sequence number 44972588
2025-08-14T02:38:33.212672Z 0 [Note] [MY-012255] [InnoDB] Removed temporary tablespace data file: "ibtmp1"
2025-08-14T02:38:33.212705Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'MEMORY'
2025-08-14T02:38:33.212717Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'CSV'
2025-08-14T02:38:33.212726Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'daemon_keyring_proxy_plugin'
2025-08-14T02:38:33.212745Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'sha2_cache_cleaner'
2025-08-14T02:38:33.212754Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'caching_sha2_password'
2025-08-14T02:38:33.212763Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'sha256_password'
2025-08-14T02:38:33.212989Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'binlog'
2025-08-14T02:38:33.213001Z 0 [Note] [MY-015020] [Server] MySQL Server: Plugins Shutdown - end.
2025-08-14T02:38:33.214872Z 0 [Note] [MY-015021] [Server] MySQL Server: Components Shutdown - start.
2025-08-14T02:38:33.215280Z 0 [Note] [MY-015022] [Server] MySQL Server: Components Shutdown - end (with return value = 0).
2025-08-14T02:38:33.215383Z 0 [System] [MY-010910] [Server] /home/shannon-bin/bin/mysqld: Shutdown complete (mysqld 8.4.3)  MySQL Community Server, and Shannon Data AI Alpha V..
2025-08-14T02:38:33.215393Z 0 [System] [MY-015016] [Server] MySQL Server - end.
2025-08-14T02:38:34.557287Z 0 [Note] [MY-013667] [Server] Error-log destination "stderr" is not a file. Can not restore error log messages from previous run.
2025-08-14T02:38:34.379980Z 0 [System] [MY-015015] [Server] MySQL Server - start.
Warning: 4T02:38:34.553588Z 0 [Warning] [MY-013711] [Server] Manifest file '/home/shannon-bin/bin/mysqld.my' is not read-only. For better security, please make sure that the file is read-only.
Warning: 4T02:38:34.553951Z 0 [Warning] [MY-011070] [Server] 'binlog_format' is deprecated and will be removed in a future release.
Warning: 4T02:38:34.554835Z 0 [Warning] [MY-010099] [Server] Insecure configuration for --secure-file-priv: Data directory is accessible through --secure-file-priv. Consider choosing a different directory.
Warning: 4T02:38:34.554845Z 0 [Warning] [MY-010101] [Server] Insecure configuration for --secure-file-priv: Location is accessible to all OS users. Consider choosing a different directory.
2025-08-14T02:38:34.554878Z 0 [Note] [MY-013932] [Server] BuildID[sha1]=9be8047c4e1d47bfcc9faec22b04e122768a9b7d
2025-08-14T02:38:34.554882Z 0 [Note] [MY-010949] [Server] Basedir set to /home/shannon-bin/.
2025-08-14T02:38:34.554884Z 0 [Note] [MY-010949] [Server] Basedir set to .
2025-08-14T02:38:34.554888Z 0 [System] [MY-010116] [Server] /home/shannon-bin/bin/mysqld (mysqld 8.4.3) starting as process 244615
2025-08-14T02:38:34.560858Z 0 [Note] [MY-015140] [InnoDB] Using innodb_flush_method=O_DIRECT
2025-08-14T02:38:34.561234Z 0 [Note] [MY-010747] [Server] Plugin 'FEDERATED' is disabled.
2025-08-14T02:38:34.561325Z 0 [Note] [MY-010747] [Server] Plugin 'ndbcluster' is disabled.
2025-08-14T02:38:34.561345Z 0 [Note] [MY-010747] [Server] Plugin 'ndbinfo' is disabled.
2025-08-14T02:38:34.561354Z 0 [Note] [MY-010747] [Server] Plugin 'ndb_transid_mysql_connection_map' is disabled.
2025-08-14T02:38:34.562561Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
2025-08-14T02:38:34.562592Z 1 [Note] [MY-013546] [InnoDB] Atomic write enabled
2025-08-14T02:38:34.562621Z 1 [Note] [MY-012932] [InnoDB] PUNCH HOLE support available
2025-08-14T02:38:34.562636Z 1 [Note] [MY-012944] [InnoDB] Uses event mutexes
2025-08-14T02:38:34.562646Z 1 [Note] [MY-012945] [InnoDB] GCC builtin __atomic_thread_fence() is used for memory barrier
2025-08-14T02:38:34.562654Z 1 [Note] [MY-012948] [InnoDB] Compressed tables use zlib 1.3.1
2025-08-14T02:38:34.564297Z 1 [Note] [MY-012951] [InnoDB] Using hardware accelerated crc32 and polynomial multiplication.
2025-08-14T02:38:34.565128Z 1 [Note] [MY-012203] [InnoDB] Directories to scan './'
2025-08-14T02:38:34.565196Z 1 [Note] [MY-012204] [InnoDB] Scanning './'
2025-08-14T02:38:34.567360Z 1 [Note] [MY-012208] [InnoDB] Completed space ID check of 8 files.
2025-08-14T02:38:34.568049Z 1 [Note] [MY-012955] [InnoDB] Initializing buffer pool, total size = 24.000000M, instances = 1, chunk size =24.000000M 
2025-08-14T02:38:34.568073Z 1 [Note] [MY-011872] [InnoDB] Setting NUMA memory policy to MPOL_INTERLEAVE
2025-08-14T02:38:34.570266Z 1 [Note] [MY-011874] [InnoDB] Setting NUMA memory policy to MPOL_DEFAULT
2025-08-14T02:38:34.570284Z 1 [Note] [MY-012957] [InnoDB] Completed initialization of buffer pool
2025-08-14T02:38:34.572680Z 0 [Note] [MY-011952] [InnoDB] If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
2025-08-14T02:38:34.573425Z 1 [Note] [MY-013532] [InnoDB] Using './#ib_16384_0.dblwr' for doublewrite
2025-08-14T02:38:34.585322Z 1 [Note] [MY-013532] [InnoDB] Using './#ib_16384_1.dblwr' for doublewrite
2025-08-14T02:38:34.619361Z 1 [Note] [MY-013566] [InnoDB] Double write buffer files: 2
2025-08-14T02:38:34.619404Z 1 [Note] [MY-013565] [InnoDB] Double write buffer pages per instance: 128
2025-08-14T02:38:34.619442Z 1 [Note] [MY-013532] [InnoDB] Using './#ib_16384_0.dblwr' for doublewrite
2025-08-14T02:38:34.619485Z 1 [Note] [MY-013532] [InnoDB] Using './#ib_16384_1.dblwr' for doublewrite
2025-08-14T02:38:34.659858Z 1 [Note] [MY-013883] [InnoDB] The latest found checkpoint is at lsn = 44923494 in redo log file ./#innodb_redo/#ib_redo137.
2025-08-14T02:38:34.659960Z 1 [Note] [MY-013086] [InnoDB] Starting to parse redo log at lsn = 44923442, whereas checkpoint_lsn = 44923494 and start_lsn = 44923392
2025-08-14T02:38:34.659982Z 1 [Note] [MY-012547] [InnoDB] Log scan progressed past the checkpoint LSN 44923442.
2025-08-14T02:38:34.659991Z 1 [Note] [MY-012551] [InnoDB] Database was not shutdown normally!
2025-08-14T02:38:34.659999Z 1 [Note] [MY-012552] [InnoDB] Starting crash recovery.
2025-08-14T02:38:34.767324Z 1 [Note] [MY-012550] [InnoDB] Doing recovery: scanned up to log sequence number 44972150
2025-08-14T02:38:34.770264Z 1 [Note] [MY-013083] [InnoDB] Log background threads are being started...
2025-08-14T02:38:34.771327Z 1 [Note] [MY-012532] [InnoDB] Applying a batch of 48 redo log records ...
2025-08-14T02:38:34.884355Z 1 [Note] [MY-013557] [InnoDB] Database page corruption or a failed file read of page [page id: space=4294967294, page number=1086]. Trying to recover it from the doublewrite file.
2025-08-14T02:38:34.884520Z 1 [Note] [MY-013550] [InnoDB] Recovered page [page id: space=4294967294, page number=1086] from the doublewrite buffer.
2025-08-14T02:38:35.067243Z 1 [Note] [MY-012533] [InnoDB] 100%
2025-08-14T02:38:35Z UTC - mysqld got signal 11 ;
Signal SIGSEGV (Address not mapped to object) at address 0xfffffffffffffffb
Most likely, you have hit a bug, but this error can also be caused by malfunctioning hardware.
BuildID[sha1]=9be8047c4e1d47bfcc9faec22b04e122768a9b7d
Thread pointer: 0x0
Attempting backtrace. You can use the following information to find out
where mysqld died. If you see no messages after this, something went
terribly wrong...
stack_bottom = 0 thread_stack 0x100000
 #0 0x55e3915425c6 _ZL18print_fatal_signaliP9siginfo_t
 #1 0x55e39154293c _Z19handle_fatal_signaliP9siginfo_tPv
 #2 0x7fa21984532f <unknown>
 #3 0x55e3910bf21e _Z23page_cur_insert_rec_lowPhP12dict_index_tPKhPmP5mtr_t.cold
 #4 0x55e392ba8e6c _Z25page_cur_parse_insert_recbPKhS0_P11buf_block_tP12dict_index_tP5mtr_t
 #5 0x55e392b658c6 _ZL32recv_parse_or_apply_log_rec_body9mlog_id_tPKhS1_jjP11buf_block_tP5mtr_tmm
 #6 0x55e392b68f8a _Z22recv_recover_page_funcbP11buf_block_t
 #7 0x55e392d68839 _Z20buf_page_io_completeP10buf_page_tbP9IORequestP10fil_node_t
 #8 0x55e392ec3a4f _Z12fil_aio_waitm
 #9 0x55e392b8ed57 _ZL17io_handler_threadm
 #10 0x55e392b8c0c5 _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJ15Detached_threadPFvmEmEEEEE6_M_runEv
 #11 0x7fa219cecdb3 <unknown>
 #12 0x7fa21989caa3 <unknown>
 #13 0x7fa219929c3b <unknown>
 #14 0xffffffffffffffff <unknown>
The manual page at http://dev.mysql.com/doc/mysql/en/crashing.html contains
information that should help you find out what is causing the crash.
Writing a core file
safe_process[244614]: Child process: 244615, killed by signal: 11
----------SERVER LOG END-------------


 - the logfile can be found in '/home/shannon-bin/mysql-test/var/log/binlog.binlog_reset_master_delete_file_failure-row/binlog_reset_master_delete_file_failure.log'

```

--- a/storage/rapid_engine/imcs/table.cpp
+++ b/storage/rapid_engine/imcs/table.cpp
@@ -121,17 +121,13 @@ RpdTable::RpdTable(const TABLE *&mysql_table, const TableConfig &config)
 }
 
 Table::Table(const TABLE *&mysql_table, const TableConfig &config) : RpdTable(mysql_table, config) {
-  // [TODO] intial global compoent.
-  // m_txn_coordinator = std::make_unique<Transaction_Coordinator>();
-  // m_version_manager = std::make_unique<Global_Version_Manager>();
-  // m_bg_workers = std::make_unique<Background_Worker_Pool>(config.background_worker_threads);
-
   // create intial IMCU
   create_initial_imcu();
 }
 
 Table::~Table() {
   {
+    std::scoped_lock lk(m_imcu_mtex);
     m_imcus.clear();
     m_imcu_index.clear();
   }

--- a/storage/rapid_engine/populate/log_redolog.cpp
+++ b/storage/rapid_engine/populate/log_redolog.cpp
@@ -137,104 +137,6 @@ bool LogParser::check_key_field(std::unordered_map<std::string, ShannonBase::key
   return false;
 }
 
-int LogParser::build_key(const Rapid_load_context *context,
-                         std::unordered_map<std::string, mysql_field_t> &field_values,
-                         std::map<std::string, key_info_t> &keys) {
-#if 0
-  std::string sch_tb = context->m_schema_name + ":" + context->m_table_name;
-  auto rpd_tb = Imcs::Imcs::instance()->get_rpd_table(sch_tb);
-  auto &matched_keys = rpd_tb->get_source_keys();
-  ut_a(matched_keys.size() > 0);
-
-  for (const auto &[key_name, key_meta] : matched_keys) {
-    const size_t key_len = key_meta.first;
-    const auto &key_fields = key_meta.second;
-
-    std::unique_ptr<uchar[]> key_buff(new uchar[key_len]);
-    memset(key_buff.get(), 0x0, key_meta.first);
-
-    uint key_offset{0u};
-    for (const std::string &key_field : key_fields) {
-      auto mysql_key = field_values.find(key_field);
-      ut_a(mysql_key != field_values.end());
-      auto &mysql_field = mysql_key->second;
-
-      if (mysql_field.has_nullbit) {
-        if (key_offset >= key_len) {
-          return HA_ERR_GENERIC;
-        }
-        *key_buff.get() = (mysql_field.is_null) ? 1 : 0;
-        key_offset++;
-      }
-
-      auto cs = rpd_tb->get_field(key_field) ? rpd_tb->get_field(key_field)->header()->m_charset : nullptr;
-
-      // blob or var string type.
-      switch (mysql_field.mtype) {
-        case DATA_BLOB:
-        case DATA_VARCHAR:
-        case DATA_VARMYSQL: {
-          if (key_offset + HA_KEY_BLOB_LENGTH + mysql_field.mlength > key_len) {
-            return HA_ERR_GENERIC;
-          }
-          int2store(key_buff.get() + key_offset, mysql_field.plength);
-          key_offset += HA_KEY_BLOB_LENGTH;
-          std::memcpy(key_buff.get() + key_offset, mysql_field.data.get(), mysql_field.mlength);
-          key_offset += mysql_field.mlength;
-        } break;
-        case DATA_DOUBLE:
-        case DATA_FLOAT:
-        case DATA_DECIMAL:
-        case DATA_FIXBINARY: {
-          if (key_offset + mysql_field.mlength > key_len) {
-            return HA_ERR_GENERIC;
-          }
-          ut_a(mysql_field.mlength == mysql_field.plength);
-          auto field_ptr = rpd_tb->get_field(key_field)->header()->m_source_fld;
-          uchar encoding[8] = {0};
-          auto val = Utils::Util::get_field_numeric<double>(field_ptr, mysql_field.data.get(), nullptr);
-          ShannonBase::Imcs::Index::Encoder<double>::EncodeData(val, encoding);
-          std::memcpy(key_buff.get() + key_offset, encoding, mysql_field.mlength);
-          key_offset += mysql_field.mlength;
-        } break;
-        case DATA_INT: {
-          if (key_offset + mysql_field.mlength > key_len) {
-            return HA_ERR_GENERIC;
-          }
-          ut_a(mysql_field.mlength == mysql_field.plength);
-          auto field_ptr = rpd_tb->get_field(key_field)->header()->m_source_fld;
-          switch (field_ptr->type()) {
-            case MYSQL_TYPE_LONG: {
-              uchar encoding[4] = {0};
-              auto val = Utils::Util::get_field_numeric<int32_t>(field_ptr, mysql_field.data.get(), nullptr);
-              ShannonBase::Imcs::Index::Encoder<int32_t>::EncodeData(val, encoding);
-              std::memcpy(key_buff.get() + key_offset, encoding, mysql_field.mlength);
-            } break;
-            default: {
-            } break;
-          }
-          key_offset += mysql_field.mlength;
-        } break;
-        default: {
-          // other types.
-          if (key_offset + mysql_field.mlength > key_len) {
-            return HA_ERR_GENERIC;
-          }
-          std::memcpy(key_buff.get() + key_offset, mysql_field.data.get(), mysql_field.mlength);
-          key_offset += mysql_field.mlength;
-          if (key_offset < key_len && cs)
-            cs->cset->fill(cs, (char *)key_buff.get() + key_offset, key_meta.first - key_offset, ' ');
-        } break;
-      }
-    }
-
-    key_info_t key_info = {key_meta.first, std::move(key_buff)};
-    keys.emplace(key_name, std::move(key_info));
-  }
-#endif
-  return ShannonBase::SHANNON_SUCCESS;
-}
-
 int LogParser::store_field_in_mysql_format(const dict_index_t *index, const dict_col_t *col, const byte *dest,
                                            const byte *src, ulint mlen, ulint plen) {
   mysql_row_templ_t templ;
@@ -523,6 +425,69 @@ byte *LogParser::parse_tablespace_redo_extend(byte *ptr, const byte *end, const 
   return ptr;
 }
 
+bool LogParser::rec_to_mysql_format(Rapid_load_context *context, mem_heap_t *heap, const rec_t *rec,
+                                    const dict_index_t *index, const ulint *offsets, uchar **mysql_rec_out,
+                                    size_t *rec_len_out) {
+  // 1. calc MySQL record length
+  size_t mysql_rec_len = 0;
+  for (ulint i = 0; i < index->n_fields; i++) {
+    const dict_col_t *col = index->get_col(i);
+    mysql_rec_len += col->get_max_size();
+    if (col->mtype == DATA_VARCHAR) mysql_rec_len += (col->len > 255) ? 2 : 1;
+  }
+
+  // NULL bitmap space.
+  mysql_rec_len += UT_BITS_IN_BYTES(index->n_nullable);
+
+  // 2. allocate MySQL format rec buffer.
+  uchar *mysql_rec = static_cast<uchar *>(mem_heap_alloc(heap, mysql_rec_len));
+  memset(mysql_rec, 0, mysql_rec_len);
+
+  // 3. create row_prebuilt_t
+  row_prebuilt_t *prebuilt = row_create_prebuilt(index->table, mysql_rec_len);
+  if (!prebuilt) return false;
+
+  // 4. convert
+  mem_heap_t *blob_heap = nullptr;
+  bool success = row_sel_store_mysql_rec(mysql_rec,  // MySQL format rec
+                                         prebuilt,   // prebuilt
+                                         rec,        // InnoDB format rec
+                                         nullptr,    // vrow
+                                         true,       // rec_clust (cluster or not)
+                                         index,      // index
+                                         index,      // prebuilt
+                                         offsets,    // offsets arrary
+                                         false,      // clust_templ_for_sec
+                                         nullptr,    // lob_undo
+                                         blob_heap   // BLOB heap
+  );
+
+  if (!success) {
+    row_prebuilt_free(prebuilt, true);
+    if (blob_heap) mem_heap_free(blob_heap);
+    return false;
+  }
+
+  // 5. get TRX_ID(if needed)
+  ulint trx_id_col = dict_table_get_sys_col_no(index->table, DATA_TRX_ID);
+  if (trx_id_col != ULINT_UNDEFINED) {
+    // get TRX_ID from MySQL rec
+    const uchar *trx_id_ptr = mysql_rec + prebuilt->mysql_prefix_len;
+    for (ulint i = 0; i < trx_id_col; i++) {
+      trx_id_ptr += prebuilt->mysql_template[i].mysql_col_len;
+    }
+    context->m_extra_info.m_trxid = mach_read_from_6(trx_id_ptr);
+  }
+
+  // 6. results
+  *mysql_rec_out = mysql_rec;
+  *rec_len_out = mysql_rec_len;
+
+  row_prebuilt_free(prebuilt, true);
+  if (blob_heap) mem_heap_free(blob_heap);
+  return true;
+}
+
 bool LogParser::rec_field_parse(Rapid_load_context *context, mem_heap_t *heap, Imcs::RpdTable *rpd_table,
                                 const rec_t *rec, const dict_index_t *index, const ulint *offsets,
                                 const dict_index_t *real_index, const ulint *real_offsets, size_t idx,
@@ -732,7 +697,6 @@ int LogParser::parse_rec_fields(Rapid_load_context *context, const rec_t *rec, c
   ut_ad(rec_offs_validate(rec, index, offsets));
   ut_ad(rec_offs_size(offsets));
 
-#if 0
   ut_a(offsets);
   ut_a(rec == nullptr || rec_get_n_fields(rec, index) >= rec_offs_n_fields(offsets));
 
@@ -741,11 +705,9 @@ int LogParser::parse_rec_fields(Rapid_load_context *context, const rec_t *rec, c
 
   auto imcs_instance = ShannonBase::Imcs::Imcs::instance();
   ut_a(imcs_instance);
-  std::string sch_tb;
-  sch_tb.append(context->m_schema_name).append(":").append(context->m_table_name);
-  auto rpd_table = imcs_instance->get_rpd_table(sch_tb);
+  auto rpd_table = dict_table_is_partition(real_index->table) ? imcs_instance->get_rpd_parttable(real_index->table->id)
+                                                              : imcs_instance->get_rpd_table(real_index->table->id);
   auto field_mutex = std::make_shared<std::mutex>();
-
   std::unique_ptr<mem_heap_t, decltype(&mem_heap_free)> heap(mem_heap_create(UNIV_PAGE_SIZE, UT_LOCATION_HERE),
                                                              mem_heap_free);
   auto heap_ptr = heap.get();
@@ -764,7 +726,6 @@ int LogParser::parse_rec_fields(Rapid_load_context *context, const rec_t *rec, c
     rec_field_parse(context, heap_ptr, rpd_table, rec, index, offsets, real_index, real_offsets, idx, *field_mutex,
                     field_values);
   }
-#endif
   return ShannonBase::SHANNON_SUCCESS;
 }
 
@@ -799,90 +760,73 @@ int LogParser::parse_cur_rec_change_apply_low(Rapid_load_context *context, const
     UNIV_MEM_ASSERT_RW(rec_start, extra_size);
   }
 #endif /* UNIV_DEBUG_VALGRIND */
-#if 0
   auto imcs_instance = ShannonBase::Imcs::Imcs::instance();
   ut_a(imcs_instance);
-  std::string sch_tb(context->m_schema_name);
-  sch_tb.append(":").append(context->m_table_name);
-  auto rpd_tb = imcs_instance->get_rpd_table(sch_tb);
-  auto log_parser_pool = ShannonBase::Imcs::Imcs::pool();
-
+  auto rpd_tb = dict_table_is_partition(real_index->table) ? imcs_instance->get_rpd_parttable(real_index->table->id)
+                                                           : imcs_instance->get_rpd_table(real_index->table->id);
   // no user defined index.
   if (!strncmp(real_index->name(), innobase_index_reserve_name, strlen(innobase_index_reserve_name))) {
     context->m_extra_info.m_key_buff = std::make_unique<uchar[]>(SHANNON_DATA_DB_ROW_ID_LEN);
     context->m_extra_info.m_key_len = SHANNON_DATA_DB_ROW_ID_LEN;
   } else {
-    auto key_len = rpd_tb->first_field()->header()->m_key_len.load();
+    auto key_len = context->m_extra_info.m_key_len;
     context->m_extra_info.m_key_buff = std::make_unique<uchar[]>(key_len);
     context->m_extra_info.m_key_len = key_len;
   }
 
+  auto res{ShannonBase::SHANNON_SUCCESS};
+  std::unique_ptr<mem_heap_t, decltype(&mem_heap_free)> heap(mem_heap_create(UNIV_PAGE_SIZE, UT_LOCATION_HERE),
+                                                             mem_heap_free);
+  auto heap_ptr = heap.get();
+  uchar *mysql_rec{nullptr};
+  size_t mysql_rec_len{0};
+  if (rec_to_mysql_format(context, heap_ptr, rec, index, offsets, &mysql_rec, &mysql_rec_len)) return HA_ERR_GENERIC;
+
   switch (type) {
-
     case MLOG_REC_DELETE: {
-      if (all) return imcs_instance->delete_rows(context, {});
+      if (all) return rpd_tb->delete_rows(context, {});
 
-      std::unordered_map<std::string, mysql_field_t> row_field_value;
-      // step 1:to get all field data of deleting row.
-      if (index->n_fields <= ShannonBase::rpd_async_column_threshold)
-        parse_rec_fields(context, rec, index, offsets, real_index, row_field_value);
-      else
-        ShannonBase::Utils::cowait_sync_with(
-            *log_parser_pool, parse_rec_fields_async(context, rec, index, offsets, real_index, row_field_value));
-
-      std::map<std::string, key_info_t> keys;
-      if (build_key(context, row_field_value, keys)) return HA_ERR_WRONG_IN_RECORD;
-
-      // step 2: go throug all the data to find the matched rows. if found the do deletion.
-      auto found_rowid = find_matched_row(context, keys);
-      if (found_rowid != INVALID_ROW_ID && imcs_instance->delete_rows(context, {found_rowid})) {
-        return HA_ERR_WRONG_IN_RECORD;
-      }
+      auto rowid = rpd_tb->locate_row(context, mysql_rec);
+      if (rowid == INVALID_ROW_ID) return HA_ERR_GENERIC;
+      res = rpd_tb->delete_row(context, rowid);
     } break;
     case MLOG_REC_INSERT: {
-      std::unordered_map<std::string, mysql_field_t> row_field_value;  //<field_name, field>
-      // step 1:to get all field data of inserting row.
-      if (index->n_fields <= ShannonBase::rpd_async_column_threshold)
-        parse_rec_fields(context, rec, index, offsets, real_index, row_field_value);
-      else
-        ShannonBase::Utils::cowait_sync_with(
-            *log_parser_pool, parse_rec_fields_async(context, rec, index, offsets, real_index, row_field_value));
-
-      // step 2: insert all field data into CUs.
-      auto rowid = 0;//imcs_instance->reserve_row_id(sch_tb);
-      if (imcs_instance->write_row_from_log(context, rowid, row_field_value)) return HA_ERR_WRONG_IN_RECORD;
-
-      // step3: insert the index info.
-      std::map<std::string, key_info_t> keys;
-      if (build_key(context, row_field_value, keys) || imcs_instance->build_indexes_from_keys(context, keys, rowid))
-        return HA_ERR_WRONG_IN_RECORD;
+      res = rpd_tb->insert_row(context, mysql_rec);
     } break;
     case MLOG_REC_UPDATE_IN_PLACE: {
       ut_a(upd);
-      std::unordered_map<std::string, mysql_field_t> row_field_value;
-      // step 1:to get all field data of updating row.
-      if (index->n_fields <= ShannonBase::rpd_async_column_threshold)
-        parse_rec_fields(context, rec, index, offsets, real_index, row_field_value);
-      else
-        ShannonBase::Utils::cowait_sync_with(
-            *log_parser_pool, parse_rec_fields_async(context, rec, index, offsets, real_index, row_field_value));
+      auto global_row_id = rpd_tb->locate_row(context, mysql_rec);
 
-      // step 2: insert all field data into CUs.
-      std::map<std::string, key_info_t> keys;
-      if (build_key(context, row_field_value, keys)) return HA_ERR_WRONG_IN_RECORD;
+      // step 1: to parse the changed fields. <changed col id, new_value>
+      auto n_cols = rpd_tb->meta().num_columns;
+      ShannonBase::Imcs::RowBuffer new_row_data(n_cols);
+      new_row_data.copy_from_mysql_fields(context, const_cast<uchar *>(mysql_rec), rpd_tb->meta().fields,
+                                          rpd_tb->meta().col_offsets.data(), rpd_tb->meta().null_byte_offsets.data(),
+                                          rpd_tb->meta().null_bitmasks.data());
 
-      // step 3: find all matched rows according to rec.
-      auto found_rowid = find_matched_row(context, keys);
-      if (found_rowid == INVALID_ROW_ID) return HA_ERR_WRONG_IN_RECORD;
+      std::unordered_map<uint32_t, ShannonBase::Imcs::RowBuffer::ColumnValue> updates;
+      for (size_t idx = 0; idx < n_cols; idx++) {
+        Field *field = rpd_tb->meta().fields[idx].source_fld;
 
-      // step 4: to update rows. TODO: update key field.
-      if (!imcs_instance->update_row_from_log(context, found_rowid, row_field_value)) return HA_ERR_WRONG_IN_RECORD;
+        ptrdiff_t offset = rpd_tb->meta().col_offsets[idx];
+        size_t field_length = field->pack_length();
+
+        // comp field is changed or not.
+        if (std::memcmp(mysql_rec + offset, mysql_rec + offset, field_length) != 0) {  // record has been changed.
+          // read the new value.
+          auto col_val = new_row_data.get_column_mutable(idx);
+          updates.emplace(idx, std::move(*col_val));
+        }
+      }
+
+      // step 2: update row.
+      res = rpd_tb->update_row(context, global_row_id, updates);
     } break;
     default:
+      assert(false);
       break;
   }
-#endif
-  return ShannonBase::SHANNON_SUCCESS;
+  return res;
 }
 
 byte *LogParser::parse_cur_and_apply_delete_mark_rec(Rapid_load_context *context, byte *ptr, /*!< in: buffer */

--- a/storage/rapid_engine/populate/log_redolog.h
+++ b/storage/rapid_engine/populate/log_redolog.h
@@ -233,13 +233,6 @@ class LogParser {
   inline bool check_key_field(std::unordered_map<std::string, ShannonBase::key_meta_t>& keys,
                               const char* field_name);
 
-  // build the keys from field values.
-  // @param field_values[in] , the map of field values. <field_name, field data>
-  // @param key_values[in/out], the map of key data, <key_name, key data>.
-  // @return 0 success, otherwise failed.
-  int build_key(const Rapid_load_context *context, std::unordered_map<std::string, mysql_field_t> &field_values,
-                 std::map<std::string, key_info_t>& keys);
-
   byte *advance_parseMetadataLog(table_id_t id, uint64_t version, byte *ptr,
                                  byte *end);
 
@@ -259,6 +252,15 @@ class LogParser {
                                                    const page_id_t &page_id,
                                                    ulint parsed_bytes,
                                                    bool parse_only);
+
+  bool rec_to_mysql_format(Rapid_load_context *context,
+                                  mem_heap_t *heap,
+                                  const rec_t *rec,
+                                  const dict_index_t *index,
+                                  const ulint *offsets,
+                                  uchar **mysql_rec_out,
+                                  size_t *rec_len_out);
+
   /** parse a rec and get all feidls data in mysql format and save these values
    * into a vector. all field values store in feild_values' and store key info
    * into context. [field_name, field_value] and [key_name, key_value]*/

--- a/storage/rapid_engine/utils/memory_pool.cpp
+++ b/storage/rapid_engine/utils/memory_pool.cpp
@@ -25,14 +25,15 @@
 */
 #include "storage/rapid_engine/utils/memory_pool.h"
 
-#include "sql/log.h"
-
 #include <algorithm>
 #include <chrono>
 #include <cstring>
 #include <iomanip>
 #include <sstream>
 #include <stdexcept>
+
+#include "mysqld_error.h"
+#include "sql/log.h"
 
 namespace ShannonBase {
 namespace Utils {
@@ -66,9 +67,6 @@ MemoryPool::SubPool::SubPool(size_t size, size_t alignment)
     : memory_base(nullptr), total_size(size), current_offset(0), is_from_parent(false), parent_ref() {
   // Allocate new memory for root pool
   memory_base = aligned_alloc_portable(alignment, size);
-  if (!memory_base) {
-    throw std::bad_alloc();
-  }
 }
 
 MemoryPool::SubPool::SubPool(void *base, size_t size)
@@ -92,14 +90,8 @@ MemoryPool::MemoryPool(const Config &config) : m_config(config), m_shutdown(fals
     initialize_pools(m_config.initial_size);
 
     // Start background monitoring thread
-    if (m_config.allow_expansion || m_config.auto_defragmentation) {
+    if (m_config.allow_expansion || m_config.auto_defragmentation)
       m_monitor_thread = std::thread(&MemoryPool::monitor_loop, this);
-    }
-
-    log(LogLevel::INFO, "MemoryPool initialized");
-    log(LogLevel::INFO, "  Total: " + format_size(m_config.initial_size));
-    log(LogLevel::INFO, "  Small Pool: " + format_size(m_subpools[0]->total_size));
-    log(LogLevel::INFO, "  Large Pool: " + format_size(m_subpools[1]->total_size));
   }
   // Sub-pool initialization is deferred to initialize_as_sub_pool()
 }
@@ -128,12 +120,8 @@ MemoryPool::~MemoryPool() noexcept {
       }
 
       // Step 2: using parent deallocate
-      try {
-        parent->deallocate(m_subpool_base, subpool_size);
-        log(LogLevel::INFO, "Successfully returned sub-pool memory block: " + format_size(subpool_size));
-      } catch (...) {
-        log(LogLevel::ERROR, "Failed to return sub-pool memory via deallocate");
-      }
+      auto res = parent->deallocate(m_subpool_base, subpool_size);
+      if (res != Result::OK) log(LogLevel::ERROR, "Failed to return sub-pool memory via deallocate");
 
       // Step 3: remove itself from parent child list
       {
@@ -154,10 +142,10 @@ MemoryPool::~MemoryPool() noexcept {
   cleanup();
 }
 
-void MemoryPool::reinitialize(const Config &new_config) {
+MemoryPool::Result MemoryPool::reinitialize(const Config &new_config) {
   if (m_config.is_sub_pool) {
-    log(LogLevel::ERROR, "reinitialize() not allowed on sub-pool");
-    return;
+    my_error(ER_SECONDARY_ENGINE_PLUGIN, MYF(0), "reinitialize() not allowed on sub-pool");
+    return MemoryPool::Result::FAILED_SUBPOOL_RECURSIVE;
   }
 
   // Step 1:reset status.
@@ -186,13 +174,9 @@ void MemoryPool::reinitialize(const Config &new_config) {
     m_monitor_thread = std::thread(&MemoryPool::monitor_loop, this);
   }
 
-  log(LogLevel::INFO, "MemoryPool reinitialized");
-  log(LogLevel::INFO, "  Total: " + format_size(m_config.initial_size));
-  log(LogLevel::INFO, "  Small Pool: " + format_size(m_subpools[0]->total_size));
-  log(LogLevel::INFO, "  Large Pool: " + format_size(m_subpools[1]->total_size));
-
   // Step 5: update statistics
   m_stats.total_capacity.store(m_config.initial_size, std::memory_order_relaxed);
+  return MemoryPool::Result::OK;
 }
 
 void *MemoryPool::allocate(size_t size, SubPoolType pool_type, const std::string &tenant_id) {
@@ -229,100 +213,64 @@ void *MemoryPool::allocate_auto(size_t size, const std::string &tenant_id) {
   return allocate(size, type, tenant_id);
 }
 
-void MemoryPool::deallocate(void *ptr, size_t size) noexcept {
-  if (!ptr) return;
+MemoryPool::Result MemoryPool::deallocate(void *ptr, size_t size) noexcept {
+  MemoryPool::Result res{MemoryPool::Result::OK};
+  if (!ptr) return res;
 
-  try {
-    std::scoped_lock lock(m_alloc_mutex);
-    auto it = m_allocations.find(ptr);
-    if (it == m_allocations.end()) {
-      log(LogLevel::WARNING, "deallocate: pointer " + std::to_string((uintptr_t)ptr) +
-                                 " not found in m_allocations, size=" + format_size(size));
-      return;
-    }
+  std::scoped_lock lock(m_alloc_mutex);
+  auto it = m_allocations.find(ptr);
+  if (it == m_allocations.end()) return res;
 
-    const AllocationInfo info = it->second;
-    auto &subpool = m_subpools[info.pool_index];
-    {
-      std::scoped_lock pool_lock(subpool->mutex);
-      FreeBlock block{info.offset, info.aligned_size};
-      subpool->free_blocks.push_back(block);
-      merge_adjacent_free_blocks(subpool.get());
-    }
-
-    if (!info.tenant_id.empty()) update_tenant_usage(info.tenant_id, -(ssize_t)size);
-
-    m_stats.used_bytes.fetch_sub(size, std::memory_order_relaxed);
-    m_stats.deallocation_count.fetch_add(1, std::memory_order_relaxed);
-    m_allocations.erase(it);
-
-    log(LogLevel::DEBUG,
-        "deallocate: freed " + format_size(info.aligned_size) + " at offset " + std::to_string(info.offset));
-
-  } catch (const std::exception &e) {
-    log(LogLevel::ERROR, "deallocate error: " + std::string(e.what()));
+  const AllocationInfo info = it->second;
+  auto &subpool = m_subpools[info.pool_index];
+  {
+    std::scoped_lock pool_lock(subpool->mutex);
+    FreeBlock block{info.offset, info.aligned_size};
+    subpool->free_blocks.push_back(block);
+    merge_adjacent_free_blocks(subpool.get());
   }
+
+  if (!info.tenant_id.empty()) update_tenant_usage(info.tenant_id, -(ssize_t)size);
+
+  m_stats.used_bytes.fetch_sub(size, std::memory_order_relaxed);
+  m_stats.deallocation_count.fetch_add(1, std::memory_order_relaxed);
+  m_allocations.erase(it);
+
+  return res;
 }
 
 bool MemoryPool::expand(size_t additional_size) {
-  if (!m_config.allow_expansion) {
-    log(LogLevel::WARNING, "Expansion not allowed");
-    return false;
+  if (!m_config.allow_expansion || m_config.is_sub_pool) return false;
+
+  if (additional_size < m_config.min_expansion_size) return false;
+
+  size_t small_pool_addition = static_cast<size_t>(additional_size * m_config.small_pool_ratio);
+  size_t large_pool_addition = additional_size - small_pool_addition;
+
+  bool success = true;
+  if (small_pool_addition > 0) success &= expand_subpool(0, small_pool_addition);
+
+  if (large_pool_addition > 0) success &= expand_subpool(1, large_pool_addition);
+
+  if (success) {
+    m_stats.expansion_count.fetch_add(1, std::memory_order_relaxed);
+    log(LogLevel::INFO, "Expanded by " + format_size(additional_size));
   }
-
-  if (m_config.is_sub_pool) {
-    log(LogLevel::WARNING, "Sub-pools cannot be expanded");
-    return false;
-  }
-
-  if (additional_size < m_config.min_expansion_size) {
-    log(LogLevel::WARNING, "Expansion size too small, minimum: " + format_size(m_config.min_expansion_size));
-    return false;
-  }
-
-  try {
-    size_t small_pool_addition = static_cast<size_t>(additional_size * m_config.small_pool_ratio);
-    size_t large_pool_addition = additional_size - small_pool_addition;
-
-    bool success = true;
-    if (small_pool_addition > 0) success &= expand_subpool(0, small_pool_addition);
-
-    if (large_pool_addition > 0) success &= expand_subpool(1, large_pool_addition);
-
-    if (success) {
-      m_stats.expansion_count.fetch_add(1, std::memory_order_relaxed);
-      log(LogLevel::INFO, "Expanded by " + format_size(additional_size));
-    }
-
-    return success;
-  } catch (const std::exception &e) {
-    log(LogLevel::ERROR, "Expansion failed: " + std::string(e.what()));
-    return false;
-  }
+  return success;
 }
 
 bool MemoryPool::defragment() {
-  if (!m_config.auto_defragmentation) {
-    log(LogLevel::WARNING, "Defragmentation not enabled");
-    return false;
+  if (!m_config.auto_defragmentation) return false;
+
+  bool success = true;
+  for (size_t i = 0; i < m_subpools.size(); ++i) success &= defragment_subpool(i);
+
+  if (success) {
+    m_stats.defragmentation_count.fetch_add(1, std::memory_order_relaxed);
+    log(LogLevel::INFO, "Defragmentation completed");
   }
 
-  try {
-    log(LogLevel::INFO, "Starting defragmentation...");
-
-    bool success = true;
-    for (size_t i = 0; i < m_subpools.size(); ++i) success &= defragment_subpool(i);
-
-    if (success) {
-      m_stats.defragmentation_count.fetch_add(1, std::memory_order_relaxed);
-      log(LogLevel::INFO, "Defragmentation completed");
-    }
-
-    return success;
-  } catch (const std::exception &e) {
-    log(LogLevel::ERROR, "Defragmentation failed: " + std::string(e.what()));
-    return false;
-  }
+  return success;
 }
 
 void MemoryPool::reset() noexcept {
@@ -334,53 +282,44 @@ void MemoryPool::reset() noexcept {
     if (!m_child_pools.empty()) return;
   }
 
-  try {
-    // reset all subpool.
-    for (auto &subpool : m_subpools) {
-      std::scoped_lock lock(subpool->mutex);
-      subpool->current_offset = 0;
-      subpool->free_blocks.clear();
-    }
-
-    // cleanup allocated track
-    {
-      std::scoped_lock lock(m_alloc_mutex);
-      m_allocations.clear();
-    }
-
-    // reset the statistics
-    {
-      std::scoped_lock lock(m_tenant_mutex);
-      for (auto &[_, tenant] : m_tenants) {
-        tenant->current_usage.store(0, std::memory_order_relaxed);
-        tenant->peak_usage.store(0, std::memory_order_relaxed);
-        tenant->allocation_count.store(0, std::memory_order_relaxed);
-      }
-    }
-
-    // reset pool statistics.
-    m_stats.allocated_bytes.store(0, std::memory_order_relaxed);
-    m_stats.used_bytes.store(0, std::memory_order_relaxed);
-    m_stats.peak_usage.store(0, std::memory_order_relaxed);
-    m_stats.allocation_count.store(0, std::memory_order_relaxed);
-    m_stats.deallocation_count.store(0, std::memory_order_relaxed);
-
-    log(LogLevel::INFO, "Reset completed");
-  } catch (const std::exception &e) {
-    log(LogLevel::ERROR, "Reset failed: " + std::string(e.what()));
+  // reset all subpool.
+  for (auto &subpool : m_subpools) {
+    std::scoped_lock lock(subpool->mutex);
+    subpool->current_offset = 0;
+    subpool->free_blocks.clear();
   }
+
+  // cleanup allocated track
+  {
+    std::scoped_lock lock(m_alloc_mutex);
+    m_allocations.clear();
+  }
+
+  // reset the statistics
+  {
+    std::scoped_lock lock(m_tenant_mutex);
+    for (auto &[_, tenant] : m_tenants) {
+      tenant->current_usage.store(0, std::memory_order_relaxed);
+      tenant->peak_usage.store(0, std::memory_order_relaxed);
+      tenant->allocation_count.store(0, std::memory_order_relaxed);
+    }
+  }
+
+  // reset pool statistics.
+  m_stats.allocated_bytes.store(0, std::memory_order_relaxed);
+  m_stats.used_bytes.store(0, std::memory_order_relaxed);
+  m_stats.peak_usage.store(0, std::memory_order_relaxed);
+  m_stats.allocation_count.store(0, std::memory_order_relaxed);
+  m_stats.deallocation_count.store(0, std::memory_order_relaxed);
 }
 
 void MemoryPool::set_tenant_quota(const std::string &tenant_id, size_t quota) {
   std::scoped_lock lock(m_tenant_mutex);
-
   auto it = m_tenants.find(tenant_id);
   if (it != m_tenants.end())
     it->second->quota = quota;
   else
     m_tenants[tenant_id] = std::make_unique<TenantConfig>(tenant_id, quota);
-
-  log(LogLevel::INFO, "Tenant '" + tenant_id + "' quota set to " + format_size(quota));
 }
 
 size_t MemoryPool::get_tenant_usage(const std::string &tenant_id) const {
@@ -392,7 +331,8 @@ size_t MemoryPool::get_tenant_usage(const std::string &tenant_id) const {
 
 std::shared_ptr<MemoryPool> MemoryPool::create_sub_pool(size_t sub_pool_size, const std::string &tenant_name) {
   if (m_config.is_sub_pool) {
-    throw std::runtime_error("Cannot create sub-pool from another sub-pool");
+    my_error(ER_SECONDARY_ENGINE_PLUGIN, MYF(0), "Cannot create sub-pool from another sub-pool");
+    return nullptr;
   }
 
   Config sub_config = m_config;
@@ -422,7 +362,8 @@ std::shared_ptr<MemoryPool> MemoryPool::create_sub_pool(size_t sub_pool_size, co
 std::shared_ptr<MemoryPool> MemoryPool::create_from_parent(const std::shared_ptr<MemoryPool> &parent_pool,
                                                            const std::string &tenant_name, size_t sub_pool_size) {
   if (!parent_pool) {
-    throw std::invalid_argument("Parent pool cannot be null");
+    my_error(ER_SECONDARY_ENGINE_PLUGIN, MYF(0), "Parent pool cannot be null");
+    return nullptr;
   }
 
   return parent_pool->create_sub_pool(sub_pool_size, tenant_name);
@@ -578,14 +519,14 @@ MemoryPool::PoolStats::Snapshot MemoryPool::stats() const noexcept { return m_st
 void MemoryPool::set_log_level(LogLevel level) { m_config.log_level = level; }
 
 bool MemoryPool::validate_config() {
-  if (m_config.initial_size < 10 * 1024 * 1024) return false;
-  // throw std::invalid_argument("Initial size must be at least 10MB");
-
-  if (m_config.small_pool_ratio < 0.0 || m_config.small_pool_ratio > 0.5) return false;
-  // throw std::invalid_argument("Small pool ratio must be between 0.0 and 0.5");
-
-  if (m_config.expansion_trigger_threshold <= 0.0 || m_config.expansion_trigger_threshold > 1.0) return false;
-  // throw std::invalid_argument("Expansion trigger must be between 0.0 and 1.0");
+  if (m_config.small_pool_ratio < 0.0 || m_config.small_pool_ratio > 0.5) {
+    my_error(ER_SECONDARY_ENGINE_PLUGIN, MYF(0), "Small pool ratio must be between 0.0 and 0.5");
+    return false;
+  }
+  if (m_config.expansion_trigger_threshold <= 0.0 || m_config.expansion_trigger_threshold > 1.0) {
+    my_error(ER_SECONDARY_ENGINE_PLUGIN, MYF(0), "Expansion trigger must be between 0.0 and 1.0");
+    return false;
+  }
   return true;
 }
 
@@ -616,33 +557,18 @@ void MemoryPool::initialize_as_sub_pool(void *parent_memory, size_t size) {
 void *MemoryPool::allocate_from_pool(int pool_idx, size_t aligned_size, size_t actual_size,
                                      const std::string &tenant_id) {
   if (unlikely(aligned_size == 0)) aligned_size = m_config.alignment;
-  if (aligned_size == 0) {
-    log(LogLevel::ERROR, "Zero alignment size after config adjustment");
-    return nullptr;
-  }
-  if (pool_idx < 0 || pool_idx >= static_cast<int>(m_subpools.size())) {
-    log(LogLevel::ERROR,
-        "Pool index out of bounds: " + std::to_string(pool_idx) + ", max=" + std::to_string(m_subpools.size()));
-    return nullptr;
-  }
+  if (aligned_size == 0) return nullptr;
+
+  if (pool_idx < 0 || pool_idx >= static_cast<int>(m_subpools.size())) return nullptr;
 
   auto &subpool = m_subpools[pool_idx];
-  if (!subpool) {
-    log(LogLevel::ERROR, "Subpool is null at index: " + std::to_string(pool_idx));
-    return nullptr;
-  }
+  if (!subpool) return nullptr;
 
   std::scoped_lock lock(subpool->mutex);
-  if (!subpool->memory_base) {
-    log(LogLevel::ERROR, "Memory base is null for subpool: " + std::to_string(pool_idx));
-    return nullptr;
-  }
+  if (!subpool->memory_base) return nullptr;
 
-  if (subpool->current_offset > subpool->total_size) {
-    log(LogLevel::ERROR, "Current offset exceeds total size: offset=" + std::to_string(subpool->current_offset) +
-                             ", total=" + std::to_string(subpool->total_size));
-    return nullptr;
-  }
+  if (subpool->current_offset > subpool->total_size) return nullptr;
+
   void *ptr = try_allocate_from_free_blocks(subpool.get(), aligned_size);
   if (ptr) {
 #ifndef NDEBUG
@@ -841,18 +767,14 @@ void MemoryPool::update_peak_usage() noexcept {
 
 void MemoryPool::monitor_loop() {
   while (!m_shutdown.load(std::memory_order_acquire)) {
-    try {
-      std::this_thread::sleep_for(std::chrono::seconds(5));
+    std::this_thread::sleep_for(std::chrono::seconds(5));
+    cleanup_expired_children();
 
-      cleanup_expired_children();
-
-      auto s = m_stats.snapshot();
-      if (m_config.allow_expansion && s.usage_percentage >= 85.0)
-        expand(std::max(m_config.initial_size / 4, m_config.min_expansion_size));
-      if (m_config.auto_defragmentation && s.fragmentation_ratio >= 0.3) {
-        defragment();
-      }
-    } catch (...) {
+    auto s = m_stats.snapshot();
+    if (m_config.allow_expansion && s.usage_percentage >= 85.0)
+      expand(std::max(m_config.initial_size / 4, m_config.min_expansion_size));
+    if (m_config.auto_defragmentation && s.fragmentation_ratio >= 0.3) {
+      defragment();
     }
   }
 }
@@ -874,10 +796,6 @@ void MemoryPool::cleanup_expired_children() noexcept {
   if (m_child_pools.size() < old_size) {
     log(LogLevel::DEBUG, "Cleaned " + std::to_string(old_size - m_child_pools.size()) + " expired sub-pools");
   }
-}
-
-void MemoryPool::log(LogLevel level, const std::string &message) const {
-  DBUG_PRINT("memory_pool", ("[%d] %s", static_cast<int>(level), message.c_str()));
 }
 
 void *MemoryPool::aligned_alloc_portable(size_t alignment, size_t size) noexcept {
@@ -914,6 +832,10 @@ std::string MemoryPool::format_size(size_t bytes) {
   std::ostringstream oss;
   oss << std::fixed << std::setprecision(2) << size << " " << units[unit];
   return oss.str();
+}
+
+void MemoryPool::log(LogLevel level, const std::string &message) const {
+  DBUG_PRINT("memory_pool", ("[%d] %s", static_cast<int>(level), message.c_str()));
 }
 }  // namespace Utils
 }  // namespace ShannonBase

--- a/storage/rapid_engine/utils/memory_pool.h
+++ b/storage/rapid_engine/utils/memory_pool.h
@@ -129,6 +129,12 @@ namespace Utils {
 class MemoryPool : public std::enable_shared_from_this<MemoryPool> {
  public:
   /**
+   * @enum Oper Result Typle
+   * @brief Type of result for allocation
+   */
+  enum class Result { OK = 0, BAD_ALLOCATION, FAILED_SUBPOOL_RECURSIVE, UNKNOWN_RESULT };
+
+  /**
    * @enum SubPoolType
    * @brief Type of sub-pool for allocation targeting
    */
@@ -240,7 +246,7 @@ class MemoryPool : public std::enable_shared_from_this<MemoryPool> {
   MemoryPool(const MemoryPool &) = delete;
   MemoryPool &operator=(const MemoryPool &) = delete;
 
-  void reinitialize(const Config &new_config);
+  Result reinitialize(const Config &new_config);
 
   // Memory Allocation/Deallocation Interface
   /**
@@ -269,7 +275,7 @@ class MemoryPool : public std::enable_shared_from_this<MemoryPool> {
    * @param size Original allocation size
    * @note This function never throws
    */
-  void deallocate(void *ptr, size_t size) noexcept;
+  Result deallocate(void *ptr, size_t size) noexcept;
 
   // Pool Management Interface
   /**

